### PR TITLE
cpu/cc2538: add periph/i2c driver

### PR DESCRIPTION
**Pull request by hexluthor**
Tested on `cc2538dk`.

### CODE REVIEW

**Comment by OlegHahm** [comment]
@hexluthor, good to see that you're still interested in RIOT and keep contributing. As you may have seen on the [mailing list](https://lists.riot-os.org/pipermail/devel/2015-September/002982.html) we're currently in feature freeze, so we will probably not be able to merge this within the next two weeks, but getting all the cc2538 stuff you have proposed is definitely a top priority right after the release.

**Comment by hexluthor** [comment]
Thanks @OlegHahm. No rush. I saw the `periph/i2c` API and wanted to get this out there.

**Comment by thomaseichinger** [comment]
@hexluthor I am sorry I didn't respond to this PR recently. I currently can't flash the OpenMote, but I am in contact with the OpenMote people to fix this again. I will test this as soon as I can.

**Comment by A-Paul** [comment]
Hi @hexluthor, I got a CC2838DK here and going to perform HW tests on that platform. For now I'm trying to get used to the components.
I tried to flash a example, following your description in the wiki, but somehow this step:
> Hold down the SELECT pushbutton while pressing RESET to activate the MCU's internal bootloader...
has not the desired effect. Invoking `make flash` gives me this error:

`ERROR: Can't connect to target. Ensure boot loader is started. (no answer to synch sequence)`

Both `/dev/ttyUSB{0,1}` are up. Do you have any hints?

**Comment by hexluthor** [comment]
@A-Paul, you're using a SmartRF06 baseboard with CC2538EM daughter board attached, correct? Make sure these jumpers are populated:

 * VDD TO EM
 * Enable UART over XDS100v3
 * VDD -> EB power

Confirm that the power switch in the on position and the USB power source is selected. Try both UARTs:
```
make PORT=/dev/ttyUSB0 flash term
make PORT=/dev/ttyUSB1 flash term
```

If you send me a photo of your board, I may be able to identify an issue. Good luck!

**Comment by kYc0o** [comment]
hello! I was wondering if I can ACK this PR using a ReMote http://zolertia.io/product/hardware/re-mote ?

**Comment by PeterKietzmann** [comment]
@kYc0o thanks for the offer! AFAIK @A-Paul already started to do hardware testing. But if you are faster ;-) ?! Generally it should be possible to test the diver on a re-mote but note that you need a logic analyser/scope and an i2c slave device to do so

**Comment by kYc0o** [comment]
@PeterKietzmann Yes I have the instruments, I'll try to test it in the following hours

**Comment by PeterKietzmann** [comment]
Nice! While you're at it how about #3842 :-) ?

**Comment by kYc0o** [comment]
Yeah sure, I already asked the people on that PR :-)

**Comment by kYc0o** [comment]
I wrote a small example to test the functionality.

The configuration is:
A cc2538dk attached to the SmartRF06 board is configured as a master i2c device and an Arduino DUE as slave. The goal is to send the message "This is the i2c working" from the master to the slave.

The cc2538dk example :

```
#include <stdio.h>
#include <string.h>
#include "periph/i2c.h"

int main(void)
{
	char *i2c_message = "This is the i2c working";
	uint8_t charLength = strlen(i2c_message);

    puts("Hello World!");

    printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
    printf("This board features a(n) %s MCU.\n", RIOT_MCU);

    if (i2c_init_master(I2C_0, I2C_SPEED_NORMAL) == 0) {
    	printf("Trying to send message \"%s\" with length %d\n", i2c_message, charLength);
    	if (i2c_write_bytes(I2C_0, 8, i2c_message, charLength) == charLength) {
    		printf("I2C message successfully sent!\n");
    	} else {
    		printf("I2C message not sent!\n");
    	}
    } else {
    	printf("Error while initializing I2C interface\n");
    }

    return 0;
}
```

The arduino sketch:

```
#include <Wire.h>

void setup() {
  Wire.begin(8);                // join i2c bus with address #8
  Wire.onReceive(receiveEvent); // register event
  Serial.begin(9600);           // start serial for output
  Serial.println("Starting I2C tests sketch!");
}

void loop() {
  delay(100);
}

// function that executes whenever data is received from master
// this function is registered as an event, see setup()
void receiveEvent(int howMany) {
  while (Wire.available()) { // loop through all available bytes
    char c = Wire.read(); // receive byte as a character
    Serial.print(c);         // print the character
  }
}
```

This is working if I plug the SmartRF06 board for the first time, then if I push the reset button to resend the message it is not sent showing the error `I2C message not sent!`. If a re-plug the board it works again. The Arduino DUE is never reset.

Any thoughts?

**Comment by PeterKietzmann** [comment]
Could you put a logic analyser in parallel to the bus to see the check the signals? I don't know how the Arduino slave implementation behaves. Maybe something with the ACK/NACK synchronization goes wrong after resetting the cc2538 board? E.g. some default configurations of the hardware pins lead to signal toggles on the SCK or SDA pins which the slave device tries to interpret as I2C command?
I usually connect to "simple" I2C hardware when testing this kind of drivers.

**Comment by PeterKietzmann** [comment]
One question: Is `8` the bus address on which the Arduino listens?

**Comment by hexluthor** [comment]
Hi all. Here's a bunch of new commits that make the I2C driver more robust. I cherry-picked them so hopefully nothing got lost. It now depends on the xtimer module, but I don't know how to express this in the Makefile rules. For now, I just add `USEMODULE += xtimer` if needed.

**Comment by kYc0o** [comment]
Ok I'll try to analyze with the oscilloscope. As I receive the message correctly on the Arduino side, I think it should behave as expected.
The address 8 is on the line:
```
Wire.begin(8);                // join i2c bus with address #8
```
I'll try to get a simpler slave i2c device too.

**Comment by kYc0o** [comment]
hi @hexluthor thanks for your commits, I'll test with and give feedback.

**Comment by PeterKietzmann** [comment]
@kYc0o any progress?

**Comment by kYc0o** [comment]
I'm waiting for an i2c hardware for more tests. I'll put the analyser output but everything seems to be ok...

**Comment by kYc0o** [comment]
@hexluthor have you tested this using tests/periph_i2c on some i2c device? I'm currently testing on cc2538dk but it freezes when I initialize the i2c

**Comment by hexluthor** [comment]
@kYc0o the xtimer module is currently broken on cpu/cc2538. I'll address that in a separate PR. For now, build without `USEMODULE += xtimer` and `init_master` should work ok. Tested on `cc2538dk`. I didn't use a scope but the `tests/periph_i2c` commands seem to work ok.

**Comment by kYc0o** [comment]
Ok right, I experienced the same problem. How did you tested without xtimer? I cannot compile it without the module...

**Comment by hexluthor** [comment]
@kYc0o, grab my latest commit e34801f which tests for `MODULE_XTIMER` within `periph/i2c.c`. You may also need to `make clean all`. I'm not sure.

**Comment by kYc0o** [comment]
I added your last commit. The commands work but there is no communication. I'll dig more on the problem and give you more details tomorrow.

**Comment by hexluthor** [comment]
Double-check the pin config. Are you using a re-mote? I'll do some more
testing today too.

**Comment by alignan** [comment]
I will test this also on the RE-Mote, I have some I2C sensors and an analyser to compare

**Comment by hexluthor** [comment]
I can display data on an I2C LCD using e34801f, `tests/periph_i2c` and `make BOARD=cc2538dk` using these commands:
```
init_master 0 1
w 120 64 48 49 50 51 52
```
Check your I2C slave address. Try multiplying or dividing by two because datasheets are sometimes ambiguous about whether the address is 7 or 8-bit. If you don't have external pull-up resistors, be sure to enable the MCU's internal pull-ups.
```
IOC_PXX_OVER[i2c_config[I2C_0].scl_pin] |= IOC_OVERRIDE_PUE;
IOC_PXX_OVER[i2c_config[I2C_0].sda_pin] |= IOC_OVERRIDE_PUE;
```
I think the internal pull-ups are on the weak side for I2C, but may work fine at lower speeds.

**Comment by kYc0o** [comment]
I made several tests, including a very simple one, without success...
as a comparison, I flashed the same code to a samr21, on which everything works as expected:
```
init_master 0 1
2016-02-03 17:14:21,600 - INFO # > init_master 0 1
2016-02-03 17:14:21,620 - INFO # I2C_0 successfully initialized as master!
w 8 2
2016-02-03 17:14:40,622 - INFO # > w 8 2
2016-02-03 17:14:40,625 - INFO # i2c_write_byte(I2C_0, 0x08, 0x02)
2016-02-03 17:14:40,652 - INFO # I2C_0: successfully wrote 1 bytes to the bus
```
This test send the hex number 0x02 through the I2C at address 8. I configured an arduino as slave on address 8 and I got the 0x02.

on a cc2538dk:
```
init_master 0 1
2016-02-03 17:05:20,874 - INFO # > init_master 0 1
2016-02-03 17:05:20,876 - INFO # I2C_0 successfully initialized as master!
w 8 2
2016-02-03 17:05:34,118 - INFO # > w 8 2
2016-02-03 17:05:34,121 - INFO # i2c_write_byte(I2C_0, 0x08, 0x02)
2016-02-03 17:05:34,129 - INFO # I2C_0: successfully wrote 0 bytes to the bus
```
for your tests
```
2016-02-03 17:39:08,612 - INFO # > init_master 0 1
2016-02-03 17:39:08,616 - INFO # I2C_0 successfully initialized as master!
w 120 64 48 49 50 51 52
2016-02-03 17:39:10,886 - INFO # > w 120 64 48 49 50 51 52
2016-02-03 17:39:10,891 - INFO # i2c_write_bytes(I2C_0, 0x78, [, 0x40, 0x30, 0x31, 0x32, 0x33, 0x34])
2016-02-03 17:39:10,895 - INFO # I2C_0: successfully wrote 0 bytes to the bus
```
and finally a remote:
```
init_master 0 1
2016-02-03 17:05:20,874 - INFO # > init_master 0 1
2016-02-03 17:05:20,876 - INFO # I2C_0 successfully initialized as master!
w 8 2
2016-02-03 17:05:34,118 - INFO # > w 8 2
2016-02-03 17:05:34,121 - INFO # i2c_write_byte(I2C_0, 0x08, 0x02)
2016-02-03 17:05:34,129 - INFO # I2C_0: successfully wrote 0 bytes to the bus
```
My analyzer does not get any changes on the signal, even when initializing the i2c interface. The arduino did't get anything as well.

Maybe I'm using the wrong pins, which are ambiguous on your code. I was wondering why the port numbers change, while the driver is being implemented for the CC2538SF53 CPU, which is exactly the same for the 3 supported platforms (cc2538dk, re-mote and openmote).

For information:
on the cc2538dk (SmartRF06) I'm using RF1.18_MOSI for SDA and RF1.16_SCL for SCL
on the remote (https://github.com/Zolertia/Resources/wiki/RE-Mote#re-mote-pin-out) PC2 for SDA and PC3 for SCL, as shown on the picture. However, on your code the pins are GPIO_PB0 for SDA and GPIO_PB1 for SCL.

I'll make more tests tomorrow to figure out what can be wrong on my side.

**Comment by hexluthor** [comment]
Hi @kYc0o. It's working for me on the same pins (SmartRF06):
```
main(): This is RIOT! (Version: 2014.12-3657-ge348-cc2538-i2c)
Test for the low-level I2C driver
> init_master 0 1
I2C_0 successfully initialized as master!
> 
> w 120 64 48 49 50 51 52
i2c_write_bytes(I2C_0, 0x78, [, 0x40, 0x30, 0x31, 0x32, 0x33, 0x34])
I2C_0: successfully wrote 6 bytes to the bus
> 
```
I wouldn't expect to see any changes on the bus during `init_master`.

I tried to select port/pin combinations that made sense for each mote PCB. In some cases SCL/SDA were already defined. In others I piggy-backed onto the SPI pins (like the SmartRF06). SPI and I2C can coexist, at least in theory. I couldn't find a proper RE-Mote pinout at the time but now that you've shown me one I will fix those pin definitions.

Without reconfiguring the Arduino slave, please try `w 16 2` and `w 4 2`. If either of those work, then I should shift the address bits to keep the behavior consistent with the samr21.

**Comment by alignan** [comment]
Just a caveat: current RE-Mote implementation in RIOT (prototype A) is not the same as the Wiki and our current platform (revision A).  @kYc0o which RE-mote are you using?

I was waiting for the feature freeze to be over to make a PR with the new updates, see:

https://github.com/alignan/RIOT/tree/pull/remote-update

**Comment by alignan** [comment]
Tested with the `bmp180` sensor (reading calibration data):

````
2016-02-04 22:54:10,626 - INFO # > rr 238 170 5
2016-02-04 22:54:10,628 - INFO # i2c_read_regs(I2C_0, 0xee, 0xaa, char *res, 5)
2016-02-04 22:54:10,629 - INFO # I2C_0: successfully read 5 bytes from reg 0xaa:
2016-02-04 22:54:10,629 - INFO #   [0x22, 0x7f, 0xfb, 0xad, 0xc6, ])
````
The sensor has the 0x77 (7-bit) I2C address, as seen above the test expects an 8-bit address

**Comment by alignan** [comment]
:+1:

**Comment by kYc0o** [comment]
I tried the same BMP180 sensor with both cc2538dk and samr21.
cc2538dk:
```
2016-02-08 18:00:43,295 - INFO # > rr 238 170 5
2016-02-08 18:00:43,298 - INFO # i2c_read_regs(I2C_0, 0xee, 0xaa, char *res, 5)
2016-02-08 18:00:43,301 - INFO # Error: no bytes were read
```
for a samr21:
```
2016-02-08 18:17:25,956 - INFO # > rr 238 208 1
2016-02-08 18:17:25,960 - INFO # i2c_read_reg(I2C_0, 0xee, 0xd0, char *res)
2016-02-08 18:17:26,011 - INFO # I2C_0: successfully read 1 bytes from reg 0xd0:
2016-02-08 18:17:26,012 - INFO #   [0x00, ])
```
However...
```
2016-02-08 18:19:48,819 - INFO # > rr 119 208 1
2016-02-08 18:19:48,823 - INFO # i2c_read_reg(I2C_0, 0x77, 0xd0, char *res)
2016-02-08 18:19:48,827 - INFO # I2C_0: successfully read 1 bytes from reg 0xd0:
2016-02-08 18:19:48,828 - INFO #   [0x55, ])
```
Are we dealing with a big inconsistency on the addresses resolution?

**Comment by kYc0o** [comment]
I forgot to add your last commit, now I'm having
```
2016-02-08 18:33:04,460 - INFO # > rr 119 208 1
2016-02-08 18:33:04,465 - INFO # i2c_read_reg(I2C_0, 0x77, 0xd0, char *res)
2016-02-08 18:33:04,469 - INFO # I2C_0: successfully read 1 bytes from reg 0xd0:
2016-02-08 18:33:04,470 - INFO #   [0x55, ])
```
on the cc2538dk.

I'll make some other tests and then we can try to merge.

**Comment by kYc0o** [comment]
Ok all my tests have passed with cc2538dk. I'll try to check tomorrow with a remote and I think I can get an openmote. If everything goes ok I think this will be merged soon and we can try to get the SPI driver working too.
Thanks @hexluthor and @alignan for your help :)

**Comment by A-Paul** [comment]
Hi @kYc0o,

> I forgot to add your last commit, now I'm having
```
2016-02-08 18:33:04,460 - INFO # > rr 119 208 1

That is the same address that worked in your comment before. What about `238`

**Comment by kYc0o** [comment]
Hi @A-Paul, in fact the test on my previous comment was on a samr21, on my last comment is on the cc2538dk. By merging the last commit `238` does not work.
Something to note is that the samr21 "reads" without error but it returns nothing, while the cc2538dk returns error if there is nothing to read (reading from a wrong address or wrong register I guess), so which is the correct behaviour?

**Comment by A-Paul** [comment]
@kYc0o 
> By merging the last commit `238` does not work.

I'm confused, `238` gave an error on *cc2538dk* in https://github.com/RIOT-OS/RIOT/pull/3765#issuecomment-181486184

> which is the correct behaviour?

I would say reporting ( and detecting) an error is correct.

**Comment by kYc0o** [comment]
> I'm confused, `238` gave an error on *cc2538dk* in https://github.com/RIOT-OS/RIOT/pull/3765#issuecomment-181486184

Yes, since the commit corrects the address size from 8-bit to 7-bit, which is, afaik, more standard. Thus it tries to read to an incorrect address.

>> which is the correct behaviour?

>I would say reporting ( and detecting) an error is correct.

Ok, so it is maybe necessary to change the implementation on samr21 to return error instead of reading 0 bytes, in order to standardize the i2c implementations.

**Comment by A-Paul** [comment]
I also tried my luck with a SRF08 distance sensor. Base address is 0xe0.
Reading register 0 should return the software version.

```2016-02-09 00:38:30,588 - INFO # > rr 240 0 4
2016-02-09 00:38:30,592 - INFO # i2c_read_regs(I2C_0, 0xf0, 0x00, char *res)
2016-02-09 00:38:30,597 - INFO # I2C_0: successfully read 1 bytes from reg 0x00:
2016-02-09 00:38:30,600 - INFO #   [0x0b, ])

Without the last commit I could use 0xe0 (224) after merging it I have to use 0xf0 (240).

But I can't write to register 1 to initiate a measurement.

> 2016-02-09 00:59:49,992 - INFO # > wr 240 1 80
2016-02-09 00:59:49,995 - INFO # i2c_write_reg(I2C_0, 0xf0, 0x01, 0x50)
2016-02-09 00:59:49,998 - INFO # Error: no bytes were written

After that reading also gives an error until I do init_master

>2016-02-09 01:01:27,165 - INFO # > rr 240 0 1
2016-02-09 01:01:27,169 - INFO # i2c_read_reg(I2C_0, 0xf0, 0x00, char *res)
2016-02-09 01:01:27,175 - INFO # Error: no bytes were read
init_master 0 1
2016-02-09 01:01:30,656 - INFO # > init_master 0 1
2016-02-09 01:01:30,660 - INFO # I2C_0 successfully initialized as master!
rr 240 0 1
2016-02-09 01:01:35,283 - INFO # > rr 240 0 1
2016-02-09 01:01:35,287 - INFO # i2c_read_reg(I2C_0, 0xf0, 0x00, char *res)
2016-02-09 01:01:35,291 - INFO # I2C_0: successfully read 1 bytes from reg 0x00:
2016-02-09 01:01:35,292 - INFO #   [0x0b, ])

What speed setting are you using?

**Comment by kYc0o** [comment]
I think this is the example of a 8-bit address, which was the default size before the commit.
Maybe this can help http://riot-os.org/api/group__driver__srf08.html .
On the registers definition the base address is set to `112` which is `0x70`

**Comment by A-Paul** [comment]
I took my info from here http://www.robot-electronics.co.uk/htm/srf08tech.html. Funny enough, address 112 is also working. But writing fails either.

**Comment by kYc0o** [comment]
I think it works since `240 -> 0xF0` and if you do a left shift it becomes `112 -> 0x70` which is what the commit does.
Anyways, writing should work but maybe there are timing restrictions while writing/reading.
If you can try to reproduce this https://github.com/RIOT-OS/RIOT/blob/master/drivers/srf08/srf08.c#L94 using the periph_i2c test.

**Comment by A-Paul** [comment]
> If you can try to reproduce this https://github.com/RIOT-OS/RIOT/blob/master/drivers/srf08/srf08.c#L94 using the periph_i2c test.

I'm not sure if I got what you want to say, but that's what I did in https://github.com/RIOT-OS/RIOT/pull/3765#issuecomment-181633386.
Register 0,1 and 2 should be writable, I get error on all.

**Comment by A-Paul** [comment]
Also tried *tests/driver_srf08* but it freezes in `srf08_init()`.

**Comment by hexluthor** [comment]
My PR originally expected a left-shifted slave address. Now it shifts the address internally like the other periph/i2c drivers. So from now on use only unshifted 7-bit addresses.

Looking at the SRF08 documentation, the default address is 0xe0 = 224 but this is already left shifted so I think you should be using address 0x70 = 112 = `SRF08_DEFAULT_ADDR` and hence `wr 112 1 80`.

@A-Paul, `srf08_init()` freezes because the `xtimer` is broken. See https://github.com/RIOT-OS/RIOT/pull/4729 .

**Comment by hexluthor** [comment]
I found a bug in `i2c_write_regs()`. Working on it now....

**Comment by A-Paul** [comment]
The last commit made some progress.
```
init_master 0 2
2016-02-10 16:12:40,427 - INFO # init_master 0 2
2016-02-10 16:12:40,431 - INFO # I2C_0 successfully initialized as master!
wr 112 0 81
2016-02-10 16:12:50,608 - INFO # > wr 112 0 81
2016-02-10 16:12:50,612 - INFO # i2c_write_reg(I2C_0, 0x70, 0x00, 0x51)
2016-02-10 16:12:50,616 - INFO # I2C_0: successfully wrote 1 bytes to register 0x00
rr 112 1 5
2016-02-10 16:13:04,324 - INFO # > rr 112 1 5
2016-02-10 16:13:04,328 - INFO # i2c_read_regs(I2C_0, 0x70, 0x01, char *res, 5)
2016-02-10 16:13:04,332 - INFO # I2C_0: successfully read 5 bytes from reg 0x01:
2016-02-10 16:13:04,335 - INFO #   [0x3b, 0x00, 0x1b, 0x00, 0x58, ])
```
I can write (trigger a measurement) and get reasonable values from he registers.

Initialized and tested successful with all speeds btw. Though I have no scope here, I'm assuming the clock is *really* changing. :)

**Comment by alignan** [comment]
I too confirm this is working with the 7-bit address, writing to read temperature data from the `bmp180` sensor.

````
2016-02-17 17:53:39,726 - INFO # > wr 119 244 46
2016-02-17 17:53:39,727 - INFO # i2c_write_reg(I2C_0, 0x77, 0xf4, 0x2e)
2016-02-17 17:53:39,729 - INFO # I2C_0: successfully wrote 1 bytes to register 0xf4
rr 119 246 3
2016-02-17 17:53:42,856 - INFO # > rr 119 246 3
2016-02-17 17:53:42,858 - INFO # i2c_read_regs(I2C_0, 0x77, 0xf6, char *res, 3)
2016-02-17 17:53:42,859 - INFO # I2C_0: successfully read 3 bytes from reg 0xf6:
2016-02-17 17:53:42,860 - INFO #   [0x78, 0x8b, 0x00, ])

````

**Comment by OlegHahm** [comment]
Any news here?

**Comment by hexluthor** [comment]
Rebased onto master. I don't know of any remaining issues.

**Comment by kYc0o** [comment]
If necessary I can do a final check tomorrow, I can test at least in a cc2538dk, but possibly on the other two boards

**Comment by kYc0o** [comment]
I just checked with a cc2538dk and an mpu9150 with a great success! we need to test it in another board?

**Comment by kYc0o** [comment]
can you change this commit name: Fix i2c_write_regs() ?

Just to make travis happy :)

**Comment by kYc0o** [comment]
As #4878 was merged, @hexluthor can you rebase and change the commit message so we can merge this asap? Thanks!

**Comment by hexluthor** [comment]
Rebased and squashed. Thanks!

File: cpu/cc2538/periph/i2c.c [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+        uint_fast8_t n;
+        for (n = 0; n < try_limit; n++) {
+            if (bus_quiet()) {
+                DEBUG("%s(): SDA released after%4u SCL pulses.\n", __FUNCTION__, n);
+                break;
+            }
+
+            assert_scl();
+
+#ifdef MODULE_XTIMER
```
Comment (A-Paul) [inline cpu/cc2538/periph/i2c.c]: Is this still necessary?

File: cpu/cc2538/periph/i2c.c [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+            xtimer_usleep(scl_delay);
+#else
+            thread_yield();
+#endif
+
+            release_scl();
+
+#ifdef MODULE_XTIMER
```
Comment (A-Paul) [inline cpu/cc2538/periph/i2c.c]: Same as line 131.

File: cpu/cc2538/periph/i2c.c [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+            DEBUG("%s(): Failed to release SDA after%4u SCL pulses.\n", __FUNCTION__, n);
+        }
+    }
+
+    /* Return to hardware mode for the I2C pins */
+    gpio_hardware_control(I2C_0_SCL_PIN);
+    gpio_hardware_control(I2C_0_SDA_PIN);
+}
+
+#ifdef MODULE_XTIMER
```
Comment (A-Paul) [inline cpu/cc2538/periph/i2c.c]: Same as line 131.

File: cpu/cc2538/periph/i2c.c [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+static void callback(void *arg)
+{
+    mutex_unlock(&i2c_wait_mutex);
+}
+#endif
+
+static uint_fast8_t i2c_ctrl_blocking(uint_fast8_t flags)
+{
+#ifdef MODULE_XTIMER
```
Comment (A-Paul) [inline cpu/cc2538/periph/i2c.c]: Same as line 131.

File: cpu/cc2538/periph/i2c.c [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+    const unsigned int xtimer_timeout = 3 * (DATA_BITS + ACK_BITS) * SEC_IN_USEC / speed_hz;
+#endif
+
+    mutex_trylock(&i2c_wait_mutex);
+
+    assert(I2CM_IMR & 1);
+    i2cm_ctrl_write(flags);
+
+#ifdef MODULE_XTIMER
```
Comment (A-Paul) [inline cpu/cc2538/periph/i2c.c]: Same as line 131.

File: cpu/cc2538/periph/i2c.c [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+void cc2538_i2c_init_master(uint32_t speed_hz)
+{
+    SYS_CTRL_RCGCI2C |= 1; /**< Enable the I2C0 clock. */
+    SYS_CTRL_SCGCI2C |= 1; /**< Enable the I2C0 clock. */
+    SYS_CTRL_DCGCI2C |= 1; /**< Enable the I2C0 clock. */
+
+    /* Reset I2C peripheral */
+    SYS_CTRL_SRI2C |= 1;
+
+#ifdef MODULE_XTIMER
```
Comment (A-Paul) [inline cpu/cc2538/periph/i2c.c]: Same as line 131.

**Comment by hexluthor** [comment]
@A-Paul, it depends if we want `xtimer` to be a hard dependency of `cpu/cc2538/periph/i2c`. How would this dependency be expressed within RIOT's make system?

With a perfect I2C bus `xtimer` probably isn't needed. On an ugly bus, it may be critical.

**Comment by A-Paul** [comment]
@hexluthor, at first, sorry for the "last minute" intervention.
I stumbled upon a -now squashed- previous commit with a topic like "avoid xtimer (for now). So i thought that might be temporary changes as long as *(x)timer* is making problems. And that they can be reverted when *timer* is fixed.

**Comment by kYc0o** [comment]
AFAIK timer is working ok, I just merged xtimers several days ago, so we can say that xtimer is no more an obstacle. However, if @hexluthor wants a perfect bus, we should avoid it's use.

IMO this is a very good and useful implementation, but as @A-Paul said, the ifdef for xtimer it's not needed anymore. So, if you guys agree, we can try to merge this without those ifdefs and try to improve this in the future without any xtimers?

**Comment by A-Paul** [comment]
@kYc0o, @hexluthor I just wanted to prevent a temporary workaround being forgotten.
I understand that the ifdefs are sensible for now. I also prefer to not having hard dependencies.

From my side, ACK for this PR in the current state. @kYc0o, do you join?

**Comment by kYc0o** [comment]
I agree, although I'd prefer to remove the `ifdefs` I think this is mergeable as is. ACK from me too.

**Comment by hexluthor** [comment]
The I2C driver will work without the xtimer module. With the xtimer module, it becomes more robust and fault-tolerant. The `ifdefs` make both scenarios possible, and hence I vote to keep them.

**Comment by kYc0o** [comment]
OK great, then ACK and merge!

**Comment by LudwigKnuepfer** [comment]
Merged without CI?

**Comment by kYc0o** [comment]
CI was green when merged, since several days ago...

**Comment by kYc0o** [comment]
https://travis-ci.org/RIOT-OS/RIOT/builds/113439509

**Comment by LudwigKnuepfer** [comment]
ah sorry

**Pull request merged**